internal/promotion/runner/builtin/git_cloner.go

```diff
@@ -155,6 +159,19 @@ func (g *gitCloner) run(
 	return promotion.StepResult{Status: kargoapi.PromotionPhaseSucceeded}, nil
 }
 
+// updateGitUser updates the gitUser fields based on the provided configuration.
+func (g *gitCloner) updateGitUser(cfg builtin.GitCloneConfig) {
```

Comment:
What you're doing in this function is not concurrency safe. There could be many goroutines using this step at once. Modifying system-level `g.gitUser` as you are here will create race conditions that could lead to one use of the step using user details that were intended for a different use of the step.

Please kill off this function and do this differently, inline, on/around line 108. If the config specifies a user, build a new `git.User` from those details. If it does not, use the system-level `g.gitUser` instead.